pennylane/ops/op_math/composite.py

```diff
@@ -0,0 +1,226 @@
+# limitations under the License.
+"""
+This submodule defines a base class for composite operations.
+"""
+import abc
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Operator
+
+
+class CompositeOp(Operator, abc.ABC):
+    """A base class for operators that are composed of other operators.
+
+    Args:
+        operands: (tuple[~.operation.Operator]): a tuple of operators which will be combined.
+
+    Keyword Args:
+        do_queue (bool): determines if the operator will be queued. Default is True.
+        id (str or None): id for the operator. Default is None.
+
+    The child composite operator should define the `_name` and `_op_symbol` properties
+    during initialization and define any relevant representations, such as
+    :meth:`~.operation.Operator.matrix` and :meth:`~.operation.Operator.decomposition`.
+    """
+
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+    _op_symbol = None
+    _name = None
+
+    def __init__(
+        self, *operands: Operator, do_queue=True, id=None
+    ):  # pylint: disable=super-init-not-called
+        self._id = id
+        self.queue_idx = None
+
+        if self._name is None:
+            raise NotImplementedError("Child class must specify _name")
+        if self._op_symbol is None:
+            raise NotImplementedError("Child class must specify _op_symbol")
+        if len(operands) < 2:
+            raise ValueError(f"Require at least two operators to combine; got {len(operands)}")
+
+        self.operands = operands
+        self._wires = qml.wires.Wires.all_wires([op.wires for op in self.operands])
+        self._hash = None
+
+        if do_queue:
+            self.queue()
+
+    def __repr__(self):
+        return f" {self.op_symbol} ".join(
+            [f"({op})" if op.arithmetic_depth > 0 else f"{op}" for op in self.operands]
+        )
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.operands = tuple(s.__copy__() for s in self.operands)
+
+        for attr, value in vars(self).items():
+            if attr not in {"operands"}:
+                setattr(copied_op, attr, value)  # TODO: exclude data?
+
+        return copied_op
+
+    def __iter__(self):
+        """Return the iterator over the underlying operands."""
+        return iter(self.operands)
+
+    def __getitem__(self, idx):
+        """Return the operand at position ``idx`` of the composition."""
+        return self.operands[idx]
+
+    def __len__(self):
+        """Return the number of operators in this composite operator"""
+        return len(self.operands)
+
+    @property
+    def op_symbol(self) -> str:
+        """The symbol used when visualizing the composite operator"""
+        return self._op_symbol
+
+    @property
+    def data(self):
+        """Create data property"""
+        return [op.parameters for op in self.operands]
```

Comment:
```suggestion
        return [op.parameters for op in self]
```

Because we can do that now!

Reply:
I vaguely remember having some issues with aliasing in the data. When copying an operator, if the parameters didn't deep copy the data, then shallowing copying and manipulating the parameters changed the base objects data. 

I am pretty sure I added tests for these cases, so I think if the tests pass we should be okay 👍🏼